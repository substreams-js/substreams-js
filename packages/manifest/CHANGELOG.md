# @substreams/manifest

## 0.6.0

### Minor Changes

- [`07d9247`](https://github.com/substreams-js/substreams-js/commit/07d9247017a3b8c89f33e646274f2d45226b25f0) Thanks [@fubhy](https://github.com/fubhy)! - Updated build pipeline

### Patch Changes

- [`c873aa6`](https://github.com/substreams-js/substreams-js/commit/c873aa61933318dd01e01a4757234337bdf2ddc0) Thanks [@fubhy](https://github.com/fubhy)! - Updated dependencies

- Updated dependencies [[`07d9247`](https://github.com/substreams-js/substreams-js/commit/07d9247017a3b8c89f33e646274f2d45226b25f0), [`c873aa6`](https://github.com/substreams-js/substreams-js/commit/c873aa61933318dd01e01a4757234337bdf2ddc0)]:
  - @substreams/core@0.7.0

## 0.5.1

### Patch Changes

- [`29f9340`](https://github.com/substreams-js/substreams-js/commit/29f9340178bf9ba0e5fd343306564fc99fb3ba47) Thanks [@fubhy](https://github.com/fubhy)! - Updated dependencies and removed `expandEnv` transform

- Updated dependencies [[`29f9340`](https://github.com/substreams-js/substreams-js/commit/29f9340178bf9ba0e5fd343306564fc99fb3ba47)]:
  - @substreams/core@0.6.2

## 0.5.0

### Minor Changes

- [`a061703`](https://github.com/substreams-js/substreams-js/commit/a06170328364c9aad8137b4712da40e7f319e4ce) Thanks [@fubhy](https://github.com/fubhy)! - Fix peer dep version of `@effect/schema`

### Patch Changes

- [`e9b9bfd`](https://github.com/substreams-js/substreams-js/commit/e9b9bfd9846dc3ffbaaa5410ea8a816789ac5ccd) Thanks [@fubhy](https://github.com/fubhy)! - Updated dependencies and removed `expandEnv` transform

- Updated dependencies [[`e9b9bfd`](https://github.com/substreams-js/substreams-js/commit/e9b9bfd9846dc3ffbaaa5410ea8a816789ac5ccd)]:
  - @substreams/core@0.6.1

## 0.4.0

### Minor Changes

- [`8156c6f`](https://github.com/substreams-js/substreams-js/commit/8156c6f853c8a0367f44039bbc286563b7996609) Thanks [@fubhy](https://github.com/fubhy)! - Update dependencies

### Patch Changes

- Updated dependencies [[`8156c6f`](https://github.com/substreams-js/substreams-js/commit/8156c6f853c8a0367f44039bbc286563b7996609)]:
  - @substreams/core@0.6.0

## 0.3.1

### Patch Changes

- Updated dependencies [[`c563781`](https://github.com/substreams-js/substreams-js/commit/c563781dfd31f1b357e61e7bf3f5d489cb7d62a7)]:
  - @substreams/core@0.5.0

## 0.3.0

### Minor Changes

- [`6b5c8ac`](https://github.com/substreams-js/substreams-js/commit/6b5c8ac885ef7a6e2312bf2f2d33641c619c4ba9) Thanks [@fubhy](https://github.com/fubhy)! - Update peers

### Patch Changes

- Updated dependencies [[`6b5c8ac`](https://github.com/substreams-js/substreams-js/commit/6b5c8ac885ef7a6e2312bf2f2d33641c619c4ba9)]:
  - @substreams/core@0.4.0

## 0.2.0

### Minor Changes

- [`5ebdd6c`](https://github.com/substreams-js/substreams-js/commit/5ebdd6c3fd07fdaa252b0a41762c28fac4daeafb) Thanks [@fubhy](https://github.com/fubhy)! - Update dependencies

### Patch Changes

- Updated dependencies [[`5ebdd6c`](https://github.com/substreams-js/substreams-js/commit/5ebdd6c3fd07fdaa252b0a41762c28fac4daeafb)]:
  - @substreams/core@0.3.0

## 0.1.0

### Minor Changes

- [`f7b8aed`](https://github.com/substreams-js/substreams-js/commit/f7b8aed31721b7cbd4a497745a858136d78cf616) Thanks [@fubhy](https://github.com/fubhy)! - Update dependencies

- [`83943e3`](https://github.com/substreams-js/substreams-js/commit/83943e38a382fec08e93788a9c86aab7738a426d) Thanks [@fubhy](https://github.com/fubhy)! - Use `effect` meta package and update all other dependencies

### Patch Changes

- Updated dependencies [[`f7b8aed`](https://github.com/substreams-js/substreams-js/commit/f7b8aed31721b7cbd4a497745a858136d78cf616), [`83943e3`](https://github.com/substreams-js/substreams-js/commit/83943e38a382fec08e93788a9c86aab7738a426d)]:
  - @substreams/core@0.2.0

## 0.0.9

### Patch Changes

- [`e7dae5c`](https://github.com/substreams-js/substreams-js/commit/e7dae5c18ab8bea7d23be4448e7ec10f123e5d13) Thanks [@fubhy](https://github.com/fubhy)! - Updated dependencies

- [`470f301`](https://github.com/substreams-js/substreams-js/commit/470f3019a7d8851339efc13d7a0148d7d45a1669) Thanks [@fubhy](https://github.com/fubhy)! - Updated dependencies

- Updated dependencies [[`e7dae5c`](https://github.com/substreams-js/substreams-js/commit/e7dae5c18ab8bea7d23be4448e7ec10f123e5d13), [`470f301`](https://github.com/substreams-js/substreams-js/commit/470f3019a7d8851339efc13d7a0148d7d45a1669)]:
  - @substreams/core@0.1.19

## 0.0.8

### Patch Changes

- [`bcd9632`](https://github.com/substreams-js/substreams-js/commit/bcd9632913b331112326575b2bb8b0c36f052d96) Thanks [@fubhy](https://github.com/fubhy)! - Refactored manifest validation to `effect`

## 0.0.7

### Patch Changes

- [`2a5f917`](https://github.com/substreams-js/substreams-js/commit/2a5f917ea96291ec228af04722570bf25d810c33) Thanks [@fubhy](https://github.com/fubhy)! - Refactored react hooks

## 0.0.6

### Patch Changes

- [`f907183`](https://github.com/substreams-js/substreams-js/commit/f9071838dbff3a92ee38bf0d6492bd14fa450699) Thanks [@fubhy](https://github.com/fubhy)! - Fixed mermaid flow charts with the word `end`.

- Updated dependencies [[`5f0a52d`](https://github.com/substreams-js/substreams-js/commit/5f0a52d185f6f44adb9229c775f84e10e029ea5c), [`f907183`](https://github.com/substreams-js/substreams-js/commit/f9071838dbff3a92ee38bf0d6492bd14fa450699)]:
  - @substreams/core@0.1.17

## 0.0.5

### Patch Changes

- [`5c9c1c6`](https://github.com/substreams-js/substreams-js/commit/5c9c1c693274e0dd5bef8c95f9160adae513734b) Thanks [@fubhy](https://github.com/fubhy)! - Updated dependencies.

- Updated dependencies [[`5c9c1c6`](https://github.com/substreams-js/substreams-js/commit/5c9c1c693274e0dd5bef8c95f9160adae513734b), [`ca0c873`](https://github.com/substreams-js/substreams-js/commit/ca0c8738234c8ddf3e4741205b8b2bd7201e2850)]:
  - @substreams/core@0.1.15

## 0.0.4

### Patch Changes

- [`c49b848`](https://github.com/substreams-js/substreams-js/commit/c49b8482630aeb18cb1e626096f358f1bfb0b681) Thanks [@fubhy](https://github.com/fubhy)! - Added exports and fixed reading of system protos.

- [`5be1808`](https://github.com/substreams-js/substreams-js/commit/5be180832efe335d0b4e5964e7a802b9bec0d92f) Thanks [@fubhy](https://github.com/fubhy)! - Updated dependencies

- Updated dependencies [[`5be1808`](https://github.com/substreams-js/substreams-js/commit/5be180832efe335d0b4e5964e7a802b9bec0d92f)]:
  - @substreams/core@0.1.13

## 0.0.3

### Patch Changes

- [`a45ae45`](https://github.com/substreams-js/substreams-js/commit/a45ae453acc00157252f9ae5f4676ed7a6736c2c) Thanks [@fubhy](https://github.com/fubhy)! - Updated dependencies.

- Updated dependencies [[`6e52c58`](https://github.com/substreams-js/substreams-js/commit/6e52c58cc25f6cb7ad2a3763bfcab23c4c84b913), [`a45ae45`](https://github.com/substreams-js/substreams-js/commit/a45ae453acc00157252f9ae5f4676ed7a6736c2c)]:
  - @substreams/core@0.1.11

## 0.0.2

### Patch Changes

- [`c3e80ba`](https://github.com/substreams-js/substreams-js/commit/c3e80ba2dd92e666e525c4a8c66f9e02058af255) Thanks [@fubhy](https://github.com/fubhy)! - Added package.json metadata

- [`6020311`](https://github.com/substreams-js/substreams-js/commit/60203113f5755352dc8f59ddf0a530c0039728f3) Thanks [@fubhy](https://github.com/fubhy)! - Completed manifest parsing

- Updated dependencies [[`c3e80ba`](https://github.com/substreams-js/substreams-js/commit/c3e80ba2dd92e666e525c4a8c66f9e02058af255), [`6020311`](https://github.com/substreams-js/substreams-js/commit/60203113f5755352dc8f59ddf0a530c0039728f3)]:
  - @substreams/core@0.1.10

## 0.0.1

### Patch Changes

- [`4aa809b`](https://github.com/substreams-js/substreams-js/commit/4aa809b80ea4819e9f16eb12d95656f0dade0756) Thanks [@fubhy](https://github.com/fubhy)! - Removed `cjs` support

- [`4aa809b`](https://github.com/substreams-js/substreams-js/commit/4aa809b80ea4819e9f16eb12d95656f0dade0756) Thanks [@fubhy](https://github.com/fubhy)! - Added manifest reader

- Updated dependencies [[`4aa809b`](https://github.com/substreams-js/substreams-js/commit/4aa809b80ea4819e9f16eb12d95656f0dade0756), [`4aa809b`](https://github.com/substreams-js/substreams-js/commit/4aa809b80ea4819e9f16eb12d95656f0dade0756)]:
  - @substreams/core@0.1.9
