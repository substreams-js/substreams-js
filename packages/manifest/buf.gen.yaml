version: v1

managed:
  enabled: true

plugins:
  - plugin: buf.build/bufbuild/es
    out: src/proto
    opt: target=ts
  - plugin: buf.build/bufbuild/connect-es
    out: src/proto
    opt: target=ts
