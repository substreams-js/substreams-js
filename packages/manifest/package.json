{
  "name": "@substreams/manifest",
  "description": "",
  "version": "0.6.0",
  "license": "MIT",
  "type": "module",
  "publishConfig": {
    "access": "public",
    "directory": "dist"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/substreams-js/substreams-js.git",
    "directory": "packages/manifest"
  },
  "scripts": {
    "build": "pnpm build-esm && pnpm build-cjs && pnpm build-annotate && build-utils pack-v2 && cp system.pb dist/dist/system.pb",
    "build-esm": "tsc -b tsconfig.build.json",
    "build-cjs": "babel build/esm --plugins @babel/transform-export-namespace-from --plugins @babel/transform-modules-commonjs --out-dir build/cjs --source-maps",
    "build-annotate": "babel build --plugins annotate-pure-calls --out-dir build --source-maps"
  },
  "dependencies": {
    "@substreams/core": "workspace:*"
  },
  "peerDependencies": {
    "@bufbuild/buf": "^1.28.0",
    "@bufbuild/protobuf": "^1.4.2",
    "@effect/schema": "^0.47.6",
    "effect": "2.0.0-next.54",
    "yaml": "^2.3.4"
  },
  "devDependencies": {
    "@bufbuild/buf": "^1.28.0",
    "@bufbuild/protobuf": "^1.4.2",
    "@effect/schema": "^0.47.6",
    "@types/node": "^20.9.0",
    "effect": "2.0.0-next.54",
    "yaml": "^2.3.4"
  }
}
