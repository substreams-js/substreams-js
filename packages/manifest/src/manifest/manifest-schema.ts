import * as Schema from "@effect/schema/Schema";
import { nameRegExp, semverRegExp } from "@substreams/core";

export const BinaryTypeSchema = Schema.literal("wasm/rust-v1");
export type BinaryType = Schema.Schema.To<typeof BinaryTypeSchema>;

export const ManifestSpecVersionSchema = Schema.literal("v0.1.0");
export type ManifestSpecVersion = Schema.Schema.To<typeof ManifestSpecVersionSchema>;

export const ProtobufSchema = Schema.struct({
  files: Schema.array(Schema.string.pipe(Schema.pattern(/\.proto$/))),
  importPaths: Schema.array(Schema.string),
});
export type Protobuf = Schema.Schema.To<typeof ProtobufSchema>;

export const PackageSchema = Schema.struct({
  name: Schema.string.pipe(Schema.pattern(nameRegExp)),
  version: Schema.string.pipe(Schema.pattern(semverRegExp)),
  url: Schema.optional(Schema.string),
  doc: Schema.optional(Schema.string),
});
export type Package = Schema.Schema.To<typeof PackageSchema>;

export const InputSchema = Schema.union(
  Schema.struct({
    map: Schema.string,
  }),
  Schema.struct({
    source: Schema.string,
  }),
  Schema.struct({
    params: Schema.string,
  }),
  Schema.struct({
    store: Schema.string,
    mode: Schema.optional(Schema.literal("get", "deltas")).withDefault(() => "get"),
  }),
);
export type Input = Schema.Schema.To<typeof InputSchema>;

export const InitialBlockSchema = Schema.compose(
  Schema.union(Schema.bigint, Schema.bigintFromSelf, Schema.BigintFromNumber),
  Schema.NonNegativeBigintFromSelf,
);

export type InitialBlock = Schema.Schema.To<typeof InitialBlockSchema>;

export const StoreModuleSchema = Schema.struct({
  kind: Schema.literal("store"),
  doc: Schema.optional(Schema.string),
  binary: Schema.optional(Schema.string),
  initialBlock: Schema.optional(Schema.lazy(() => InitialBlockSchema)),
  inputs: Schema.array(Schema.lazy(() => InputSchema)),
  name: Schema.string.pipe(Schema.pattern(nameRegExp)),
  valueType: Schema.union(
    Schema.literal("bytes", "string", "int64", "float64", "bigint", "bigfloat", "bigdecimal"),
    Schema.templateLiteral(Schema.literal("proto:"), Schema.string),
  ),
  updatePolicy: Schema.literal("set", "set_if_not_exists", "append", "add", "min", "max"),
}).pipe(
  Schema.filter((value) => {
    const combinations: string[] = [
      "max:bigint",
      "max:int64",
      "max:bigdecimal",
      "max:bigfloat",
      "max:float64",
      "min:bigint",
      "min:int64",
      "min:bigdecimal",
      "min:bigfloat",
      "min:float64",
      "add:bigint",
      "add:int64",
      "add:bigdecimal",
      "add:bigfloat",
      "add:float64",
      "set:bytes",
      "set:string",
      "set:proto",
      "set:bigdecimal",
      "set:bigfloat",
      "set:bigint",
      "set:int64",
      "set:float64",
      "set_if_not_exists:bytes",
      "set_if_not_exists:string",
      "set_if_not_exists:proto",
      "set_if_not_exists:bigdecimal",
      "set_if_not_exists:bigfloat",
      "set_if_not_exists:bigint",
      "set_if_not_exists:int64",
      "set_if_not_exists:float64",
      "append:bytes",
      "append:string",
    ];

    const valueType = value.valueType.startsWith("proto:") ? "proto" : value.valueType;
    const combination = `${value.updatePolicy}:${valueType}`;
    return combinations.includes(combination);
  }),
);
export type StoreModule = Schema.Schema.To<typeof StoreModuleSchema>;

export const MapModuleSchema = Schema.struct({
  kind: Schema.literal("map"),
  doc: Schema.optional(Schema.string),
  binary: Schema.optional(Schema.string),
  initialBlock: Schema.optional(Schema.lazy(() => InitialBlockSchema)),
  inputs: Schema.array(Schema.lazy(() => InputSchema)).pipe(Schema.minItems(1)),
  name: Schema.string.pipe(Schema.pattern(nameRegExp)),
  output: Schema.struct({
    type: Schema.string,
  }),
});
export type MapModule = Schema.Schema.To<typeof MapModuleSchema>;

export const ModuleSchema = Schema.union(StoreModuleSchema, MapModuleSchema);
export type Module = Schema.Schema.To<typeof ModuleSchema>;

export const BinarySchema = Schema.struct({
  file: Schema.string,
  type: Schema.lazy(() => BinaryTypeSchema),
  native: Schema.optional(Schema.string),
  content: Schema.optional(Schema.instanceOf(Uint8Array)),
  entrypoint: Schema.optional(Schema.string),
  protoPackageMapping: Schema.optional(Schema.record(Schema.string, Schema.string)),
});
export type Binary = Schema.Schema.To<typeof BinarySchema>;

export const SinkSchema = Schema.struct({
  type: Schema.string,
  module: Schema.string,
  config: Schema.any,
});
export type Sink = Schema.Schema.To<typeof SinkSchema>;

export const ManifestSchema = Schema.struct({
  network: Schema.optional(Schema.string),
  specVersion: Schema.lazy(() => ManifestSpecVersionSchema),
  binaries: Schema.optional(
    Schema.record(
      Schema.string,
      Schema.lazy(() => BinarySchema),
    ),
  ),
  imports: Schema.optional(
    Schema.record(Schema.string.pipe(Schema.pattern(/^[A-Za-z_][A-Za-z0-9_-]*$/)), Schema.string),
  ),
  modules: Schema.array(Schema.lazy(() => ModuleSchema)).pipe(Schema.minItems(1)),
  package: Schema.lazy(() => PackageSchema),
  protobuf: Schema.lazy(() => ProtobufSchema),
  params: Schema.optional(Schema.record(Schema.string, Schema.string)),
  sink: Schema.optional(Schema.lazy(() => SinkSchema)),
});
export type Manifest = Schema.Schema.To<typeof ManifestSchema>;

export function parseManifestJson(input: unknown): Manifest {
  const parse = Schema.parseSync(ManifestSchema);
  return parse(input, { errors: "all" });
}
