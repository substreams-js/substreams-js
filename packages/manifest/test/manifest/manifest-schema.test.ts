import * as fs from "node:fs";
import * as path from "node:path";
import * as url from "node:url";
import * as Schema from "@effect/schema/Schema";
import { InitialBlockSchema, parseManifestJson } from "@substreams/manifest/manifest/manifest-schema";
import { Option } from "effect";
import { expect, test } from "vitest";
import { parse as parseYaml } from "yaml";

function readManifest(name: string): unknown {
  const file = url.fileURLToPath(new URL(`./__fixtures__/${name}.yaml`, import.meta.url));
  const contents = fs.readFileSync(file, "utf-8");
  return Object.assign(parseYaml(contents), {
    workDir: path.dirname(file),
  });
}

test("can validate the block index manifest", () => {
  const manifest = readManifest("block-index");
  expect(parseManifestJson(manifest)).toMatchSnapshot();
});

test("can validate the block filter manifest", () => {
  const manifest = readManifest("block-filter");
  expect(parseManifestJson(manifest)).toMatchSnapshot();
});

test("can validate the uniswap v3 manifest", () => {
  const manifest = readManifest("uniswap-v3");
  expect(parseManifestJson(manifest)).toMatchSnapshot();
});

test("can validate the ethereum block metadata manifest", () => {
  const manifest = readManifest("eth-block-metadata");
  expect(parseManifestJson(manifest)).toMatchSnapshot();
});

test("allows different initial block formats", () => {
  const parse = Schema.decodeOption(InitialBlockSchema);
  expect(parse("123")).toMatchObject(Option.some(123n));
  expect(parse(461246n)).toMatchObject(Option.some(461246n));
  expect(parse(0)).toMatchObject(Option.some(0n));
});

test("does not allow negative initial block values", () => {
  const parse = Schema.decodeOption(InitialBlockSchema);
  expect(parse("-12345")).toMatchObject(Option.none());
  expect(parse(-12345n)).toMatchObject(Option.none());
  expect(parse(-12345)).toMatchObject(Option.none());
});
