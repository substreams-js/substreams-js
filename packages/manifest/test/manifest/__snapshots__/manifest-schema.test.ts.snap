// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`can validate the block filter manifest 1`] = `
{
  "binaries": {
    "default": {
      "file": "target/wasm32-unknown-unknown/release/injective_common.wasm",
      "type": "wasm/rust-v1",
    },
  },
  "modules": [
    {
      "blockFilter": {
        "module": "index_events",
        "query": {
          "params": true,
        },
      },
      "doc": "\`filtered_events\` reads from \`all_events\` and applies a filter on the event types and attribute keys,
only outputing the events that match the filter.
Example usage: \`(type:message && attr:action) || (type:wasm && attr:_contract_address)\`
",
      "inputs": [
        {
          "params": "string",
        },
        {
          "map": "all_events",
        },
      ],
      "kind": "map",
      "name": "filtered_events",
      "output": {
        "type": "proto:sf.substreams.cosmos.v1.EventList",
      },
    },
  ],
  "package": {
    "name": "injective_common",
    "version": "v0.2.2",
  },
  "params": {
    "filtered_events": "(type:message && attr:action) || (type:wasm && attr:_contract_address)",
  },
  "specVersion": "v0.1.0",
}
`;

exports[`can validate the block index manifest 1`] = `
{
  "binaries": {
    "default": {
      "file": "target/wasm32-unknown-unknown/release/injective_common.wasm",
      "type": "wasm/rust-v1",
    },
  },
  "modules": [
    {
      "inputs": [
        {
          "map": "all_events",
        },
      ],
      "kind": "blockIndex",
      "name": "index_events",
      "output": {
        "type": "proto:sf.substreams.index.v1.Keys",
      },
    },
  ],
  "package": {
    "name": "injective_common",
    "version": "v0.2.2",
  },
  "specVersion": "v0.1.0",
}
`;

exports[`can validate the ethereum block metadata manifest 1`] = `
{
  "binaries": {
    "default": {
      "file": "./target/wasm32-unknown-unknown/release/substreams.wasm",
      "type": "wasm/rust-v1",
    },
  },
  "imports": {
    "database_change": "https://github.com/streamingfast/substreams-sink-database-changes/releases/download/v1.2.0/substreams-database-change-v1.2.0.spkg",
    "entities_change": "https://github.com/streamingfast/substreams-sink-entity-changes/releases/download/v1.3.0/substreams-sink-entity-changes-v1.3.0.spkg",
    "kv_operations": "https://github.com/streamingfast/substreams-sink-kv/releases/download/v2.1.5/substreams-sink-kv-v2.1.5.spkg",
  },
  "modules": [
    {
      "inputs": [
        {
          "source": "sf.ethereum.type.v2.Block",
        },
      ],
      "kind": "store",
      "name": "store_block_meta_start",
      "updatePolicy": "set_if_not_exists",
      "valueType": "proto:eth.block_meta.v1.BlockMeta",
    },
    {
      "inputs": [
        {
          "source": "sf.ethereum.type.v2.Block",
        },
      ],
      "kind": "store",
      "name": "store_block_meta_end",
      "updatePolicy": "set",
      "valueType": "proto:eth.block_meta.v1.BlockMeta",
    },
    {
      "inputs": [
        {
          "mode": "deltas",
          "store": "store_block_meta_start",
        },
        {
          "mode": "deltas",
          "store": "store_block_meta_end",
        },
      ],
      "kind": "map",
      "name": "db_out",
      "output": {
        "type": "proto:sf.substreams.sink.database.v1.DatabaseChanges",
      },
    },
    {
      "inputs": [
        {
          "mode": "deltas",
          "store": "store_block_meta_start",
        },
        {
          "mode": "deltas",
          "store": "store_block_meta_end",
        },
      ],
      "kind": "map",
      "name": "kv_out",
      "output": {
        "type": "proto:sf.substreams.sink.kv.v1.KVOperations",
      },
    },
    {
      "inputs": [
        {
          "mode": "deltas",
          "store": "store_block_meta_start",
        },
        {
          "mode": "deltas",
          "store": "store_block_meta_end",
        },
      ],
      "kind": "map",
      "name": "graph_out",
      "output": {
        "type": "proto:sf.substreams.sink.entity.v1.EntityChanges",
      },
    },
  ],
  "package": {
    "name": "substreams_eth_block_meta",
    "version": "v0.4.3",
  },
  "protobuf": {
    "files": [
      "block_meta.proto",
    ],
    "importPaths": [
      "./proto",
    ],
  },
  "specVersion": "v0.1.0",
}
`;

exports[`can validate the uniswap v3 manifest 1`] = `
{
  "binaries": {
    "default": {
      "file": "target/wasm32-unknown-unknown/release/substreams_uniswap_v3.wasm",
      "type": "wasm/rust-v1",
    },
  },
  "imports": {
    "entity": "https://github.com/streamingfast/substreams-entity-change/releases/download/v1.1.0/substreams-entity-change-v1.1.0.spkg",
  },
  "modules": [
    {
      "doc": "Dynamic data sources pattern for Uniswap V3 pools. This module will loop over block transactions and detect
pools created events. Once the pool created events have been detected, we will create \`Pools\` structs.

Try with
\`\`\`
substreams gui substreams.yaml map_pools_created -e $SUBSTREAMS_ENDPOINT -t +1000
\`\`\`
",
      "initialBlock": 12369621n,
      "inputs": [
        {
          "source": "sf.ethereum.type.v2.Block",
        },
      ],
      "kind": "map",
      "name": "map_pools_created",
      "output": {
        "type": "proto:uniswap.types.v1.Pools",
      },
    },
    {
      "inputs": [
        {
          "map": "map_pools_created",
        },
      ],
      "kind": "store",
      "name": "store_pools_created",
      "updatePolicy": "set",
      "valueType": "proto:uniswap.types.v1.Pool",
    },
    {
      "inputs": [
        {
          "map": "map_pools_created",
        },
      ],
      "kind": "store",
      "name": "store_tokens",
      "updatePolicy": "add",
      "valueType": "int64",
    },
    {
      "inputs": [
        {
          "map": "map_pools_created",
        },
      ],
      "kind": "store",
      "name": "store_pool_count",
      "updatePolicy": "add",
      "valueType": "bigint",
    },
    {
      "initialBlock": 12369621n,
      "inputs": [
        {
          "map": "map_pools_created",
        },
      ],
      "kind": "map",
      "name": "map_tokens_whitelist_pools",
      "output": {
        "type": "proto:uniswap.types.v1.ERC20Tokens",
      },
    },
    {
      "inputs": [
        {
          "map": "map_tokens_whitelist_pools",
        },
      ],
      "kind": "store",
      "name": "store_tokens_whitelist_pools",
      "updatePolicy": "append",
      "valueType": "string",
    },
    {
      "initialBlock": 12369621n,
      "inputs": [
        {
          "source": "sf.ethereum.type.v2.Block",
        },
        {
          "mode": "get",
          "store": "store_pools_created",
        },
      ],
      "kind": "map",
      "name": "map_extract_data_types",
      "output": {
        "type": "proto:uniswap.types.v1.Events",
      },
    },
    {
      "inputs": [
        {
          "map": "map_extract_data_types",
        },
      ],
      "kind": "store",
      "name": "store_pool_sqrt_price",
      "updatePolicy": "set",
      "valueType": "proto:uniswap.types.v1.Events.PoolSqrtPrice",
    },
    {
      "initialBlock": 12369621n,
      "inputs": [
        {
          "source": "sf.substreams.v1.Clock",
        },
        {
          "map": "map_extract_data_types",
        },
        {
          "mode": "get",
          "store": "store_pools_created",
        },
      ],
      "kind": "store",
      "name": "store_prices",
      "updatePolicy": "set",
      "valueType": "bigdecimal",
    },
    {
      "initialBlock": 12369621n,
      "inputs": [
        {
          "source": "sf.substreams.v1.Clock",
        },
        {
          "map": "map_extract_data_types",
        },
      ],
      "kind": "store",
      "name": "store_pool_liquidities",
      "updatePolicy": "set",
      "valueType": "bigint",
    },
    {
      "inputs": [
        {
          "source": "sf.substreams.v1.Clock",
        },
        {
          "map": "map_extract_data_types",
        },
      ],
      "kind": "store",
      "name": "store_total_tx_counts",
      "updatePolicy": "add",
      "valueType": "bigint",
    },
    {
      "inputs": [
        {
          "source": "sf.substreams.v1.Clock",
        },
        {
          "map": "map_extract_data_types",
        },
        {
          "mode": "get",
          "store": "store_pools_created",
        },
        {
          "mode": "get",
          "store": "store_total_tx_counts",
        },
        {
          "mode": "get",
          "store": "store_eth_prices",
        },
      ],
      "kind": "store",
      "name": "store_swaps_volume",
      "updatePolicy": "add",
      "valueType": "bigdecimal",
    },
    {
      "inputs": [
        {
          "map": "map_extract_data_types",
        },
      ],
      "kind": "store",
      "name": "store_native_amounts",
      "updatePolicy": "set",
      "valueType": "bigdecimal",
    },
    {
      "initialBlock": 12369621n,
      "inputs": [
        {
          "source": "sf.substreams.v1.Clock",
        },
        {
          "map": "map_extract_data_types",
        },
        {
          "mode": "get",
          "store": "store_pools_created",
        },
        {
          "mode": "get",
          "store": "store_prices",
        },
        {
          "mode": "get",
          "store": "store_tokens_whitelist_pools",
        },
        {
          "mode": "get",
          "store": "store_native_amounts",
        },
        {
          "mode": "get",
          "store": "store_pool_liquidities",
        },
      ],
      "kind": "store",
      "name": "store_eth_prices",
      "updatePolicy": "set",
      "valueType": "bigdecimal",
    },
    {
      "initialBlock": 12369621n,
      "inputs": [
        {
          "map": "map_extract_data_types",
        },
      ],
      "kind": "store",
      "name": "store_token_tvl",
      "updatePolicy": "add",
      "valueType": "bigdecimal",
    },
    {
      "initialBlock": 12369621n,
      "inputs": [
        {
          "source": "sf.substreams.v1.Clock",
        },
        {
          "map": "map_extract_data_types",
        },
        {
          "mode": "get",
          "store": "store_token_tvl",
        },
        {
          "mode": "get",
          "store": "store_pools_created",
        },
        {
          "mode": "get",
          "store": "store_eth_prices",
        },
      ],
      "kind": "store",
      "name": "store_derived_tvl",
      "updatePolicy": "set",
      "valueType": "bigdecimal",
    },
    {
      "initialBlock": 12369621n,
      "inputs": [
        {
          "source": "sf.substreams.v1.Clock",
        },
        {
          "mode": "deltas",
          "store": "store_derived_tvl",
        },
      ],
      "kind": "store",
      "name": "store_derived_factory_tvl",
      "updatePolicy": "add",
      "valueType": "bigdecimal",
    },
    {
      "doc": "### Hello world

This is how you use this:

\`\`\`
substreams gui blah blah
\`\`\`
",
      "inputs": [
        {
          "source": "sf.substreams.v1.Clock",
        },
        {
          "map": "map_extract_data_types",
        },
      ],
      "kind": "store",
      "name": "store_ticks_liquidities",
      "updatePolicy": "add",
      "valueType": "bigint",
    },
    {
      "inputs": [
        {
          "map": "map_extract_data_types",
        },
      ],
      "kind": "store",
      "name": "store_positions",
      "updatePolicy": "set",
      "valueType": "proto:uniswap.types.v1.Events.PositionEvent",
    },
    {
      "inputs": [
        {
          "source": "sf.substreams.v1.Clock",
        },
        {
          "mode": "deltas",
          "store": "store_prices",
        },
        {
          "mode": "deltas",
          "store": "store_eth_prices",
        },
      ],
      "kind": "store",
      "name": "store_min_windows",
      "updatePolicy": "min",
      "valueType": "bigdecimal",
    },
    {
      "inputs": [
        {
          "source": "sf.substreams.v1.Clock",
        },
        {
          "mode": "deltas",
          "store": "store_prices",
        },
        {
          "mode": "deltas",
          "store": "store_eth_prices",
        },
      ],
      "kind": "store",
      "name": "store_max_windows",
      "updatePolicy": "max",
      "valueType": "bigdecimal",
    },
    {
      "initialBlock": 12369621n,
      "inputs": [
        {
          "source": "sf.substreams.v1.Clock",
        },
        {
          "mode": "deltas",
          "store": "store_pool_count",
        },
        {
          "mode": "deltas",
          "store": "store_total_tx_counts",
        },
        {
          "mode": "deltas",
          "store": "store_swaps_volume",
        },
        {
          "mode": "deltas",
          "store": "store_derived_factory_tvl",
        },
        {
          "mode": "deltas",
          "store": "store_eth_prices",
        },
        {
          "map": "map_extract_data_types",
        },
        {
          "map": "map_pools_created",
        },
        {
          "mode": "deltas",
          "store": "store_pool_sqrt_price",
        },
        {
          "mode": "get",
          "store": "store_pool_sqrt_price",
        },
        {
          "mode": "deltas",
          "store": "store_pool_liquidities",
        },
        {
          "mode": "deltas",
          "store": "store_token_tvl",
        },
        {
          "mode": "deltas",
          "store": "store_prices",
        },
        {
          "mode": "get",
          "store": "store_prices",
        },
        {
          "mode": "get",
          "store": "store_tokens",
        },
        {
          "mode": "deltas",
          "store": "store_tokens_whitelist_pools",
        },
        {
          "mode": "deltas",
          "store": "store_derived_tvl",
        },
        {
          "mode": "deltas",
          "store": "store_ticks_liquidities",
        },
        {
          "mode": "get",
          "store": "store_total_tx_counts",
        },
        {
          "mode": "get",
          "store": "store_eth_prices",
        },
        {
          "mode": "get",
          "store": "store_positions",
        },
        {
          "mode": "deltas",
          "store": "store_min_windows",
        },
        {
          "mode": "deltas",
          "store": "store_max_windows",
        },
      ],
      "kind": "map",
      "name": "graph_out",
      "output": {
        "type": "proto:sf.substreams.entity.v1.EntityChanges",
      },
    },
  ],
  "package": {
    "doc": "These substream modules compose all the things to build up the Uniswap v3 substreams.
See individual modules docs for help.
",
    "name": "uniswap_v3",
    "url": "https://github.com/streamingfast/substreams-uniswap-v3",
    "version": "v0.2.8",
  },
  "protobuf": {
    "files": [
      "uniswap/v1/uniswap.proto",
    ],
    "importPaths": [
      "./proto",
    ],
  },
  "specVersion": "v0.1.0",
}
`;
